Py3.8: Transparently use a Python virtual environment under the hood

michaelDCurran opened the pull request:
### Link to issue number:
Fixes #11056 
Fixes #11794 
Fixes #12059 
Fixes #12060 
Fixes #12061 
Fixes #12062 
Fixes #12063 
Fixes #12065 
Fixes #12116 

### Summary of the issue:
NVDA and its build system have has many Python dependencies, the majority of which are fetched as git submodules to official git repositories, some are git submodules for our own git repositories containing pre-built dependencies (E.g. `wxPython`) and some are fetched via `pip`. And even the ones via `pip` are fetched differ as to where they are fetched into our build system.
 
The current approach to Python dependencies as the following issues:
* There is no standard way of tracking / fetching Python dependencies
* For certain dependencies such as `wxPython`, there is a great deal of extra work to pre-build and store in a git repository each time we want to upgrade to a new version of that dependency
* Python dependencies installed within our build system may conflict with Python packages already installed globally on the developer's system
* Python dependencies for NVDA may be inappropriately installed globally on the developer's system, which could affect things outside of NVDA.

As we move to `Python 3.8`, some of our Python dependencies (namely `wxPython`, `pySerial` and `py2exe`) need to be upgraded. We should take this opportunity to standardize on a cleaner way of fetching Python dependencies that address all the above issues. 

### Description of how this pull request fixes the issue:
A Python virtual environment is now used transparently by the NVDA build system, and all Python dependencies are installed into this environment using pip.
From a developer's perspective using the NVDA build system, there is no need to worry about the Python virtual environment behind the scenes, nor should you create and or activate one manually. All NVDA build system commands will handle this transparently.
* To build NVDA, SCons should continue to be used in the usual way. E.g. executing scons.bat in the root of the repository. Running py -m SCons is no longer supported, and scons.py has also been removed.
* To run NVDA from source, rather than executing source/nvda.pyw directly, the developer should now use runnvda.bat in the root of the repository. If you do try to execute source/nva.pyw, a message box will alert you this is no longer supported.
* To perform unit tests, execute rununittests.bat [<extra unittest discover options>]
* To perform system tests: execute runsystemtests.bat [<extra robot options>]
* To perform linting, execute runlint.bat <base branch>
Behind the scenes, the above batch files (scons, runnvda, rununittests, runsystemtests and runlint) ensure that the Python virtual environment is created and up to date, activates the environment, runs the command and then deactivates. All transparently. A developer should not have to know about the Python virtual environment at all.
The first time one of these commands are run, the virtual environment will be created, and all required Python dependencies will be installed with pip. You can see the entire list of packages and their exact versions that pip will use, in requirements.txt in the root of the repository.
venvUtils/ensureVenv.py contains the logic to check for, create and update the virtual environment.
If a previous virtual environment exists but has a miss-matching Python version or pip package requirements have changed, The virtual environment is recreated with the updated version of Python and packages.
If a virtual environment is found but does not seem to be ours, the user is asked if it should be overwritten or not.
This script also detects if it is being run from an existing 3rd party Python virtual environment and aborts if so. thus, it is impossible to execute SCons or NVDA from source within another Python virtual environment.
venvUtils/ensureAndActivate.bat can be used to ensure the virtual environment exists and is up to date, and then activates it in the current shell, ready for other commands to be executed in the context of NVDA's build system Python virtual environment. this would never normally be executed by itself, though appVeyor uses it at the beginning of its execution and leaves the environment active for the remainder of its execution.
venvUtils/venvCmd.bat is a script that runs a single command within the context of the NVDA build system Python virtual environment. It ensures and activates the environment, executes the command, and then deactivates the environment. this script is what all the high-level batch files use internally.
SConstruct, and source/nvda.pyw both contain logic that detects the NVDA build system Python virtual environment, and abort if it is not active.
As this PR is specific for Python 3.8, it also upgrades the following Python dependencies:
* wxPython to 4.1.1
* pySerial to 3.5
* py2exe to 0.10.1.0

### Testing strategy:
* Compiled NVDA locally with `scons source`, `scons dist` and `scons launcher`
* Run unit tests and system tests locally with no errors.
* Successfully generated try builds on appVeyor.

### Known issues with pull request:
None so far.

### Change log entry:
Changes for developers:
```
- NVDA now requires Python 3.8.
- NVDA's build system now fetches all Python dependencies with pip and stores them in a Python virtual environment. This is all done transparently.
* To build NVDA, SCons should continue to be used in the usual way. E.g. executing scons.bat in the root of the repository. Running ``py -m SCons``  is no longer supported, and ``scons.py`` has also been removed. 
* To run NVDA from source, rather than executing ``source/nvda.pyw`` directly, the developer should now use ``runnvda.bat`` in the root of the repository. If you do try to execute ``source/nvda.pyw``, a message box will alert you this is no longer supported.
* To perform unit tests, execute ``rununittests.bat [<extra unittest discover options>]``
* To perform system tests: execute ``runsystemtests.bat [<extra robot options>] ``
* To perform linting, execute ``runlint.bat <base branch>``
```

### Code Review Checklist:

This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review and confirm you have considered the following items.

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

josephsl committed:
Build NVDA with Python 3.8 32-bit (#12042)

* SConstruct: specify Pyhton 3.8.

* SConstruct: remove Python 3.7.6 check.

The latest version of Python 3.7 is 3.7.9 (as of February 2021). Therefore remove 3.7.6 check.

* SConstruct: update copyright year

* SCons.bat: run SCons under Python 3.8 (32-bit).

* Revert "SConstruct: update copyright year"

This reverts commit 096f61b1a3507c4ecc8a1ae5b289c5862c9a9962 due to impending Py2Exe upgrade that will update copyright header.

* scons.bat: update comment

* Readme: mention Python 3.8.

* appveyor.yml: Instruct appveyor to use Python 3.8

Co-authored-by: Michael Curran <mick@nvaccess.org>

michaelDCurran committed:
Remove any git submodules to distributions we can instead fetch with pip.

michaelDCurran committed:
List all required Python packages in a requirements.txt in the root of the repository.
Remove the existing requirements.txt files and logic for devDocs and system tests and linting.
Have appveyor fetch all Python dependencies at the beginning of the build.

michaelDCurran committed:
Ssconscripts: import SCons correctly.

michaelDCurran committed:
Appveyor: Create and activate a Python virtual environment. Also use scons.exe rather than our scons.py
Remove scons.bat and scons.py.

michaelDCurran committed:
Update readme.md

lukaszgo1 commented on appveyor.yml:
```diff
@@ -59,6 +59,16 @@ init:
+ # Ensure we have the latest version of pip
+ - py -m pip install --upgrade pip
+ # Create a Python virtual environment
+ - py -3.8-32 -m venv .venv
```
Since `PY_PYTHON` is being set to 3.8 x32 above you can just use `py` here.

lukaszgo1 commented on requirements.txt:
```diff
@@ -0,0 +1,26 @@
+# NVDA's runtime dependencies
+comtypes==1.1.8
+pyserial==3.5
+wxPython==4.1.1
```
Installing wx like that will also pull NumPy from PyPI. As mentioned by @josephsl in #11967 if NumPy is installed Py2exe bundles parts of in in NVDA's binary builds because ComTypes has an optional dependency on NumPy arrays. NumPy is also a dependency of screencaplibrary. Perhaps we need to be able to specify `--no-deps` for a particular package?

feerrenrut replied:
NumPy will be excluded from the Py2exe bundle explicitly, see `source/setup.py`. I think NumPy is required indirectly by robotframework-screencaplibrary anyway.

lukaszgo1 reviewed (commented):
- Please add .venv to the .gitignore.
- Why exactly you've deleted `scons.bat` and `sconns.py` when we would look into automating this process they probably would need to be reintroduced in some form.

michaelDCurran committed:
appveyor.yml: no need to specifify python version in specific commands as it is set globally.

michaelDCurran committed:
.gitignore: ignore the .venv directory

michaelDCurran committed:
setup.py: don't package numpy with NVDA.

michaelDCurran committed:
Use py2exe 0.10.1.0 rather than py2exe 0.10.2.0 which seems to actually import some of our modules and then break because of globalVars.appDir not existing yet.

josephsl committed:
Python 3.8/liblouis helper: add NVDA executable path by calling os.add_dll_directory (#12020)

* Louis helper: update copyright header

* Louis helper/Python 3.8: call os.add_dll_directory when importing libluis DLL.

Python 3.8 adds os.add_dll_directory to securely load DLL's from known locations, which affects Louis helper when trying to load liblouis.dll from NVDA executable folder. Therefore add NVDA executable path to known DLL directories through os.add_dll_directory, which resolves ModuleNotFoundError.

* Louis helper: remove header comment, use globalVars.appDir as executable path.

Reviewed by Lukasz Golonka: remove file name from header comment. Also, because __file__ will point to library.zip, use globalVars.appDir to locate NVDA executable where liblouis.dll actually resides.

* Louis helper: use globalVars.appDir directly rather than going through os.path.dirname.

Pointed out by Lukasz Golonka: globalVars.appDir records the actual path to the executable, including in source code version (nvda.pyw). Therefore use app dir directly.

michaelDCurran committed:
Merge branch 'py3.8' into py3.8_manualVenv

michaelDCurran committed:
Use comtypes 1.1.7.

michaelDCurran committed:
System tests: don't specify Python 3.7 when running system tests with source version of NVDA

michaelDCurran committed:
appveyor: run nose with the --traverse-namespace option as it is now required for locating unit tests in Python 3.8.

michaelDCurran committed:
Update to miscDeps containing brlAPI for Python 3.8

michaelDCurran committed:
Work around for issue #12058: In wxPython >= 4.1, wx.CallAfter no longer executes callbacks while NVDA's main thread is within apopup menu or message box. (#12072)

Monkeypatch wx.CallAfter to post a WM_NULL message to our top-level window after calling the original CallAfter, which causes wx's event loop to wake up enough to execute the callback.

michaelDCurran committed:
Merge remote-tracking branch 'origin/py3.8' into py3.8_manualVenv

michaelDCurran committed:
Explicitly include xmlrpc.server with py2exe, as it is required for system tests by robotremoteserver, but it can't seem to be imported from our NVDA profile directory.

MarcoZehe commented on readme.md:
````diff
@@ -119,14 +112,30 @@ Although this [must be run manually](#linting-your-changes), developers may wish
+To activate the virtual environment:
+```
+.venv\scripts\activate.bat
+```
````
Actually, I would switch the paragraphs around and simply instruct users to first create a virtual environment and then run the pip command. Since this is what AppVeyor will also do, and it is good practice anyway, why bother with howevers and maybes? Just my 2 Cents.

feerrenrut replied:
I agree. I think we should move all the "doing a local build" stuff into a dedicated file or at least section. It should be ordered so it can be followed step by step.

michaelDCurran committed:
sourceEnv: stop the user from running with no virtual environment.

michaelDCurran committed:
Integrate the Python virtual environment transparently into the build system.
It is now possible to run the scons command the same way you normally would, with out worrying about having to create or activate a virtual environment. This is now done behind the scenes for you.
Note that you must execute scons, or scons.bat though, not py -m SCons or some other invocation.
To run NVDA from source, it is now necessary to execute runnvda (runnvda.bat) in the root of the repository. This script will block until NVDA exits, though you can still run it in the background by doing
```
start /b runnvda
```
Note that trying to execute source\nvda.pyw directly will now  display an error as the NVDA build system Python virtual environment will not be detected doing it this way.
Behind the scenes, SCons and runnvda ensure that the Python virtual environment is created and up to date, activates the environment, runs the command and then deactivates. All transparently. A developer should not have to know about the Python virtual environment at all.

venvUtils/ensureVenv.py contains the logic to check for, create and update the virtual environment.
If a previous virtual environment exists but has a miss-matching Python version or pip package requirements have changed, The virtual environement is recreated with the updated version of Python and packages.
If a virtual environement is found but does not seem to be ours, the user is asked if it should be overwritten or not.
This script also detects if it is being run from an existing 3rd party Python virtual environment and aborts if so. thus, it is impossible to execute SScons or NVDA from source within another Python virtual environment.

venvUtils/ensureAndActivate.bat can be used to ensure the virtual environment exists and is up to date, and then activates it in the current shell, ready for other commands to be executed in the context of NVDA's build system Python virtual environment. this would never nroamlly be executed by itself, though appveyor uses it at the beginning of its execution and leaves the environment active for the remainder of its execution.

venvUtils/venvCmd.bat is a script that runs a single command within the context of the NVDA build system Python virtual environment. It ensures and activates the environment, executes the command, and then deactivates the environment. this script is what scons.bat and runnvda.bat use internally.

SConstruct, and source/nvda.pyw both contain logic that detects the NVDA build system Python virtual environment, and abort if it is not active.

michaelDCurran committed:
Fix typo

michaelDCurran committed:
ensureVenv.py: flush standard out before running any sub processes otherwise appveyor renders some messages in the wrong order.

lukaszgo1 commented on appveyor.yml:
```diff
@@ -59,6 +59,12 @@ init:
 
 install:
+ # Ensure we have the latest version of pip
+ - py -m pip install --upgrade pip
```
Upgrading `pip` here does not prevent warnings about it being outdated during packages installation because when creating an venv a predefined version of `pip` is used rather than one currently installed. It makes sense to update it after our venv is activated.

lukaszgo1 commented:
Also for  whatever reason NVDA fails to exit properly after install on AppVeyor with the following log:

```

CRITICAL - watchdog._crashHandler (10:38:49.715) - MainThread (4732):
NVDA crashed! Minidump written to C:\projects\nvda\testOutput\nvda_install.log\..\nvda_crash.dmp
INFO - watchdog._crashHandler (10:38:49.715) - MainThread (4732):
Listing stacks for Python threads:
Python stack for thread 5020 (watchdog):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 932, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "watchdog.pyc", line 112, in _watcher
  File "winKernel.pyc", line 228, in waitForSingleObject

Python stack for thread 5056 (winInputHook):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 932, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "winInputHook.pyc", line 79, in hookThreadFunc

Python stack for thread 3044 (_UIAHandler.UIAHandler.MTAThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 932, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "_UIAHandler.pyc", line 329, in MTAThreadFunc
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 302, in wait

Python stack for thread 1068 (braille._BgThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 932, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "braille.pyc", line 2193, in func

Python stack for thread 4732 (MainThread):
  File "nvda.pyw", line 261, in <module>
  File "core.pyc", line 576, in main
  File "wx\core.pyc", line 2237, in MainLoop
  File "watchdog.pyc", line 213, in _crashHandler
  File "watchdog.pyc", line 63, in getFormattedStacksForAllThreads

INFO - watchdog._crashHandler (10:38:49.715) - MainThread (4732):
Restarting due to crash

```

michaelDCurran committed:
remove some unneeded imports.

michaelDCurran committed:
Update pip within the virtual environment, not before, as a stock-standard pip gets copied in to the environment even if it were updated beforehand.

lukaszgo1 commented on source/core.py:
```diff
@@ -560,6 +560,15 @@ def run(self):
+	# NVDA may reconnect to read it later,
+	# but it is better to assume we are not connected to anything at the start.
+	try:
+		ctypes.windll.kernel32.FreeConsole()
```
This code assumes that `FreeConsole` would raise a `WindowsError` if  NVDA wasn't connected to the console but it seems not to be the case. According to the documentation this function returns true if the console has been fried and False if `FreeConsole` fails or there was no connected console in the first place.

lukaszgo1 commented on tests/system/libraries/NvdaLib.py:
```diff
@@ -51,7 +51,7 @@ def __init__(self):
 		if self.whichNVDA == "source":
 			self._runNVDAFilePath = _pJoin(self.repoRoot, "source/nvda.pyw")
-			self.baseNVDACommandline = f"pyw -3.7-32 {self._runNVDAFilePath}"
+			self.baseNVDACommandline = f"pyw {self._runNVDAFilePath}"
```
Can't you just use runnvda.bat here?

feerrenrut replied:
Yes, running the bat file might be better. It would be good to centralize the method used to invoke python. I've run into nasty issues before where a different version of python was used to execute the system tests. Hence the `-3.7-32` argument.

lukaszgo1 replied:
I don't think you've addressed this one.

lukaszgo1 commented on venvUtils/ensureVenv.py:
```diff
@@ -0,0 +1,129 @@
+	print("Creating virtual environment...", flush=True)
+	subprocess.run(
+		[
+			"py", "-3.8-32",
```
Specifying Python version explicitly in more places creates more places in which we would need to update it in the future. You can just use `sys.executable` here.

lukaszgo1 commented on venvUtils/ensureVenv.py:
```diff
@@ -0,0 +1,129 @@
+	"""
+	Ensures that the NVDA build system's Python virtual environment is created and up to date.
+	If a previous virtual environment exists but has a miss-matching Python version or pip package requirements have changed,
+	The virtual environement is recreated with the updated version of Python and packages.
```
environement -> environment

lukaszgo1 commented on venvUtils/ensureVenv.py:
```diff
@@ -0,0 +1,129 @@
+	Ensures that the NVDA build system's Python virtual environment is created and up to date.
+	If a previous virtual environment exists but has a miss-matching Python version or pip package requirements have changed,
+	The virtual environement is recreated with the updated version of Python and packages.
+	If a virtual environement is found but does not seem to be ours,
```
Also here...

lukaszgo1 commented on venvUtils/ensureVenv.py:
```diff
@@ -0,0 +1,129 @@
+		or not os.path.exists(venv_orig_requirements_path)
+	):
+		if askYesNoQuestion(
+			f"Virtual environement at {venv_path} probably not created by NVDA. This directory must be removed before continuing. Should it be removed?"
```
And here - probably bad copy/paste or autocomplete.

lukaszgo1 reviewed (commented):
Please make sure that readme in tests/system reflects the new situation.

michaelDCurran committed:
ensureVenv: address review commnts.

michaelDCurran committed:
core.main: FreeConsole returns 0 on failure rather than raising an exception.

michaelDCurran committed:
Added runnvdaw.bat which runs NVDA without blocking. this provides the behaviour which developers were previously used to with pythonw.

michaelDCurran commented on runnvda.bat:
```diff
@@ -0,0 +1,2 @@
+@echo off
+call "%~dp0\venvUtils\venvCmd.bat" py "%~dp0\source\nvda.pyw" %*
```
If called from a batch file, pythonw will actually block. Calls in batch files always wait on any process executed. And since it blocks, and I added the freeConsole call in core.py so that the console is still usable by NVDA, we may as well use python rather than pythonw so that critical errors are not missed.
However, I have now also added runnvdaw.bat which executes NvDA in the virtual environment with start pyw, which again provides the behaviour developers were used to by using pyw or executing nvda.pyw directly from Explorer etc. I.e. it does not block, and no console window stays around for the life time of its execution.
Of course this being a draft pr, all this is still up for debate -- I certainly welcome other ideas on the best way we can allow developers to run NVDA.

lukaszgo1 commented on appveyor.yml:
```diff
@@ -59,6 +59,10 @@ init:
+ # Activate the NvDA build system Python virtual environment, creating / updating it if necessary.
+ - .\venvUtils\ensureAndActivate.bat
+ # Save an exact list of the actual Python packages and their versions that got installed, to aide in reproducing a build
+ - py -m pip freeze >output\pip_packages.txt
```
At this point there is no directory called 'output' - it is created later on by SCons, so this command fails.

lukaszgo1 commented on venvUtils/ensureVenv.py:
```diff
@@ -0,0 +1,129 @@
+
+if __name__ == '__main__':
+	# Ensure we are not inside an already active Python virtual environment. 
+	virtualEnv = os.getenv("VIRTUAL_ENV")
```
Ideally this should not warn if the active venv is the NVDA venv.

michaelDCurran committed:
Merge branch 'master' into py3.8

michaelDCurran committed:
Merge branch 'py3.8' into py3.8_manualVenv

michaelDCurran committed:
appveyor:  push the list of installed python packages directly as an artifact when created.

michaelDCurran committed:
linting for venvUtils/ensureVenv.py

michaelDCurran committed:
Linting for source/setup.py

michaelDCurran committed:
add rununittests.bat to provide a faster way of running the unit tests without having to execute SCons (which can take a long time to load).
Also stop the error in source/garbageHandler.py seen at the end of unittest runs as Python is shutting down.

michaelDCurran committed:
Add runsystemtests.bat as a quick way of running the system tests, without having to use SCons which takes a long time to load.

michaelDCurran committed:
Add runlint.bat which provides a quicker way of linting. Rather than needing to execute SCons which can take a long time to load.

dingpengyu commented:
Using Windows version 10.0.19042 workstation testing NVDA version try-py3.8_venv-21803



Successfully tested in Simplified Chinese
[nvda.log](https://github.com/nvaccess/nvda/files/6066209/nvda.log)

michaelDCurran commented:
I believe I have addressed all review comments up to now.
I have now updated the pr description, and this pr is no longer a draft.
I'm looking for final reviews before this can possibly be merged.

codeofdusk commented on requirements.txt:
```diff
@@ -0,0 +1,26 @@
+# NVDA's build system is SCons
+SCons==4.1.0.post1
+
+# NVDA's runtime dependencies
```
Please also add `diff_match_patch_python` so that DMP will work again on Python 3.8 and newer Python versions going forward (currently the extension in the submodule is only compatible with 3.7). Once we switch to 3.8 I'll remove the extension from the nvda_dmp repo.

codeofdusk replied:
I've now opened PR #12113 (to be merged into this branch and fix DMP).

michaelDCurran committed:
Add diff-match-patch as a Python package requirement as NVDA_DMP needs it.

codeofdusk committed:
Sync with latest nvda_dmp. (#12113)

leonardder commented on requirements.txt:
```diff
@@ -0,0 +1,29 @@
+# NVDA's runtime dependencies
+comtypes==1.1.7
+pyserial==3.5
+wxPython==4.1.1
```
Please consider adding configobj from git:

```suggestion
wxPython==4.1.1
git+git://github.com/DiffSK/configobj@3e2f4cc#egg=configobj
```

Note, this is the last commit from current master, therefore this would require a readme update. No major changes

michaelDCurran committed:
Cnfigobj is now fetched by pip, rather than it being a git submodule.

feerrenrut commented on appveyor.yml:
```diff
@@ -95,12 +100,12 @@ build_script:
+ - scons source %sconsArgs%
  # We don't need launcher to run checkPot, so run the checkPot before launcher.
- - py scons.py checkPot %sconsArgs%
+ - scons checkPot %sconsArgs%
```
It might be better to run the same commands that developers are expected to run here. The same thought applies to manually activating the virtual environment and installing dependencies.

feerrenrut commented on scons.bat:
```diff
@@ -1,13 +1,3 @@
-	"%~dp0\scons.py" %*
-)
+rem Executes SScons within the NVDA build system's Python virtual environment.
+"%~dp0\venvUtils\venvCmd.bat" py -m SCons %*
```
Please add a new line to the end of this file.

feerrenrut commented on sconstruct:
```diff
@@ -7,6 +7,16 @@ import sys
+virtualEnv = os.getenv("VIRTUAL_ENV")
+if not nvdaVenv or nvdaVenv != virtualEnv:
+	print(
+		"Error: SCons cannot detect  the NVDA build system Python virtual environment.\n"
```
Double space after 'detect'

feerrenrut commented on source/core.py:
```diff
@@ -560,6 +560,11 @@ def run(self):
 	else:
 		log.debug("initializing updateCheck")
 		updateCheck.initialize()
+	# If running from source, try to disconnect from the console we may have been executed in.
+	# NVDA may reconnect to read it later,
+	# but it is better to assume we are not connected to anything at the start.
+	if ctypes.windll.kernel32.FreeConsole() != 0:
+		log.debug("Freed from parent console process.")
```
Why is this change necessary or related to the rest of the PR?

michaelDCurran replied:
If running NvDA from source in a console with runnvda.bat (in other words python) (not runnvdaw.bat or pythonw) NVDA is associated with the controle, and standard out / standard error are printed there. However, as a screen reader, NVDA itself connects to the console with focus in order to access its contents (at least the older implementation that doesn't use UIA). However, as NVDA is already connected to the console when it starts, then the winConsoleHandler code fails to connect again and or manage connecting / disconnecting from the console with focus. Thus, once NVDA is fully started up, and any critical errors would have been written to the console, as a screen reader, it is better to forcefully disconnect, and it may or may not reconnect later in order to read its content. However, I would be happy to remove this commit and address it in a future pr as it is not specifically to do with Python 3.8. However developers will still hit this if executing runnvda.bat.

feerrenrut commented on source/garbageHandler.py:
```diff
@@ -20,7 +20,9 @@ class TrackedObject:
 
 	def __del__(self):
-		notifyObjectDeletion(self)
+		# notifyObjectDeletion could be None if Python is shutting down.
+		if notifyObjectDeletion:
+			notifyObjectDeletion(self)
```
Could you elaborate on this some more? Has this behavior changed with Python 3.8?

michaelDCurran replied:
this has not changed with Python 3.8. In fact you reported this bug to me many months ago. I only really noticed it properly as I was testing the rununittests.bat file. I can remove this commit and address it in a future pr if you like? Though now making unittests much easier and faster to run with the new command, that traceback is going to become more noticeable.

feerrenrut commented on venvUtils/ensureAndActivate.bat:
```diff
@@ -0,0 +1,9 @@
+py -3.8-32 "%~dp0\ensureVenv.py"
+if ERRORLEVEL 1 exit /b %ERRORLEVEL%
+call "%~dp0\..\.venv\scripts\activate.bat"
+set NVDA_VENV=%VIRTUAL_ENV%
```
Please add a new line to the end of this file.

feerrenrut commented on venvUtils/ensureVenv.py:
```diff
@@ -0,0 +1,137 @@
+	print("Installing packages in virtual environment...", flush=True)
+	subprocess.run(
+		[
+			# Activate virtual environment
+			os.path.join(venv_path, "scripts", "activate.bat"),
+			"&&",
+			# Ensure we have the latest version of pip
+			"py", "-m", "pip",
+			"install", "--upgrade", "pip",
+			"&&",
+			# Install required packages with pip
+			"py", "-m", "pip",
+			"install", "-r", requirements_path,
```
This should probably check that each or the operations were successful, and perhaps the output should be logged somewhere?

michaelDCurran replied:
Because of the "&&"s and the check=True, if any of those commands fail with an error code, the whole command will fail I believe.
I don't feel it is really that necessary to save that output to a file -- we don't tend to do that for scons or other commands (yet). Perhaps that is something that we can look at in future?

feerrenrut commented on venvUtils/venvCmd.bat:
```diff
@@ -0,0 +1,11 @@
+call "%~dp0\ensureAndActivate.bat"
+call %*
+call "%~dp0\..\.venv\scripts\deactivate.bat"
+endlocal
```
Please add a newline to the end of this file.

feerrenrut commented on source/nvda.pyw:
```diff
@@ -18,14 +19,27 @@ if getattr(sys, "frozen", None):
+	# Ensure we are inside the NVDA build system's Python virtual environment.
+	nvdaVenv = os.getenv("NVDA_VENV")
+	virtualEnv = os.getenv("VIRTUAL_ENV")
+	if not nvdaVenv or nvdaVenv != virtualEnv:
```
This will make it difficult to do the virtual env setup manually, though perhaps this should be considered beyond scope. I was thinking that setting up or modifying the virtual env manually might be helpful when experimenting with new libraries, avoiding recreating the virtual env each time. On the other hand, recreating the virtual env each time from the requirements file will prevent silly mistakes that lead to drawing incorrect conclusions while experimenting.

michaelDCurran replied:
The only way we could allow this would be that if nvda.pyw is executed outside of the virtual environment that we then execute a subprocess of runnvda or runnvdaw.bat depending on whether python or pythonw was used to originally execute it. But I think that starts getting into some possible recursion danger. I feel it is probably better and safer not to allow executing like this from source anymore, without using runnvda or runnvdaw.

feerrenrut reviewed (changes requested):
Overall this looks really good!!

I'd also like to see this documented in the devDocs folder, I commented on the docs issue: https://github.com/nvaccess/nvda/issues/12059#issuecomment-789513379

michaelDCurran committed:
Address review comments: final line ending / extra spaces.

michaelDCurran commented on source/nvda.pyw:
```diff
@@ -10,6 +10,7 @@
 import sys
 import os
 import globalVars
+import ctypes
```
The only way we could allow this would be that if nvda.pyw is executed outside of the virtual environment that we then execute a subprocess of runnvda or runnvdaw.bat depending on whether python or pythonw was used to originally execute it. But I think that starts getting into some possible recursion danger. I feel it is probably better and safer not to allow executing like this from source anymore, without using runnvda or runnvdaw.

michaelDCurran committed:
venvCmd.bat: allow executing a command when already inside an NVDA Python virtual environment.
This therefore allows running any of the high-level batch scripts (scons.bat, runnvda.bat etc) after manually ensuring and activating the NVDA python virtual environment with venvUtils\ensureAndActivate.bat.
Developers thus now can either just run the high-level commands (letting everything  happen transparently), Or they can ensure and activate the environment once with venvUtils\ensureAndActivate.bat, and then execute anything within the context of the environment, which may include those high-level commands. Note though this second way of doing this is still not the recommended way for most developers.

michaelDCurran committed:
appveyor: abort on failure of scons and other commands.

michaelDCurran committed:
Ensure that scons.bat and other high-level commands correctly propagate their error codes.
Specifically:
* scons.bat should execute venvCmd.bat with call, like all of the other high-level batch scripts.
* On errors in batch scripts, use goto :EOF  rather than exit /b %ERRORLEVEL% (which does not always seem to work)
* venvCmd.bat: abort early if there was an error ensuring and activating the virtual environment.

michaelDCurran committed:
Revert "core.main: FreeConsole returns 0 on failure rather than raising an exception."

This reverts commit 2a9df2635472a5363a43d5bed8373bbeb3d8855b.

michaelDCurran committed:
core.main: revert addition of FreeConsole. This will be handled in a separate pr.

michaelDCurran committed:
For now runnvda.bat (python) has been removed, and runnvdaw.bat (pythonw) has been renamed to runnvda.bat. Thus, the recommended (and only) way to launch NVDA from source now is via runnvda.bat, which will not block. this is most similar to the original way of pythonw source/nvda.pyw.
We may re-introduce another batch file perhaps called runnvda_console.bat in the future, but only when the console association issue is addressed in a separate pr.

michaelDCurran committed:
garbageHandler: revert fix to "Nonetype is not callable" when Python is shutting down, as this will be addressed in a separate pr.

michaelDCurran committed:
Remove targets for unit tests, system tests and linting from SCons as they all now have their own run* commands in the root of the repository which are much faster to execute.

michaelDCurran committed:
Appveyor now calls runlint, rununittests and runsystemtests rather than using its own code.

michaelDCurran committed:
Update system tests readme to talk about runsystemtests, and not SCons.

michaelDCurran committed:
readme: briefly mention Nose and Robot and providing them extra arguments via the rununittests and runsystemtests scripts respectively.

michaelDCurran committed:
tests/sconscript: remove all code unrelated to checkPot.

michaelDCurran committed:
appveyor: call our batch files as a relative path otherise Powershell cannot find them.

michaelDCurran committed:
appveyor: py3.8 is a branch we should run builds for.

michaelDCurran committed:
runlint.bat: fix path.

michaelDCurran committed:
add a description of the build system to the devDocs directory.

michaelDCurran committed:
NVDA build system commands no longer refuse to run if not in NVDA's specific build system Python virtual environment, rather any detected Python virtual environment is now allowed. However, a warning is issued that this is not recommended, and when running NVDA from source, a warning is also logged.
Developers should still continue to run the NVDA build system commands outside of any existing Python virtual environment, however it is allowed so as to support scenarios such as launching from within an IDE that supports Python virtual environments.

michaelDCurran committed:
Linting

feerrenrut commented on appveyor.yml:
```diff
@@ -61,9 +62,12 @@ clone_depth: 1
  - py -m pip freeze >installed_python_packages.txt
+ - if ERRORLEVEL 1 exit %ERRORLEVEL%
  - appveyor PushArtifact  installed_python_packages.txt
+ - if ERRORLEVEL 1 exit %ERRORLEVEL%
```
Rather than do this here, it could be done in the `on_finish` section. Docs:
>Finalize both successful and failed builds:

It would need to first call `.\venvUtils\ensureAndActivate.bat`

Then, this will be testing the dev workflow more closely. EG calling `scons source` sets up the venv and activates.

feerrenrut replied:
@michaelDCurran What do you think of this suggestion?

feerrenrut commented on devDocs/buildSystemNotes.md:
```diff
@@ -0,0 +1,23 @@
+
+From a developer's perspective using the NVDA build system, there is no need to worry about the Python virtual environment behind the scenes, nor should you create and or activate one manually. All NVDA build system commands will handle this transparently.
+
+* To build NVDA, SCons should continue to be used in the usual way. E.g. executing scons.bat in the root of the repository. Running `py -m SCons`  is no longer supported, and scons.py has also been removed. 
```
I don't think we need to go over how to build, run, or execute tests here. Those should be documented in a single place, currently that is the main readme. I'd make this doc more of a how / why explanation.

I'd certainly references to unsupported "old ways" of doing this. The change log / PR can used to see what has changed. The python file is no longer available, so it would be pretty hard to make that mistake.

I think we should add an explicit line that builds / running should be done from a cmd.exe window, but that probably belongs in the main readme.md file.

After the first little introduction about how "the build system uses virtual environments transparently", I'd just have a couple of headings.
- How the build system works
  This should highlight the names of the main entry point scripts, list all scripts with an overview of their purpose, and certainly point out they all rely on `ensureVenvAndActivate`.
- Motivation for using virtual environments
  Eg In order to ensure the build environment is clean, no conflicts with other installed packages

michaelDCurran replied:
Perhaps you can have a go at re-wording / re-structuring this doc if you have some good ideas in your head already. Though I have taken all your other suggestions.

feerrenrut replied:
Yep, I'll take a look at this.

feerrenrut commented on devDocs/buildSystemNotes.md:
```diff
@@ -0,0 +1,23 @@
+A Python virtual environment is now used transparently by the NVDA build system, and all Python dependencies are installed into this environment using `pip`.
```
Rather than "now" I'd make this a statement about the current behavior, it is more simple to read the concrete current truth rather than interpreting a series of changes.

```suggestion
A Python virtual environment is used transparently by the NVDA build system, and all Python dependencies are installed into this environment using `pip`.

```

feerrenrut commented on devDocs/buildSystemNotes.md:
```diff
@@ -0,0 +1,23 @@
+A Python virtual environment is now used transparently by the NVDA build system, and all Python dependencies are installed into this environment using `pip`.
+
+From a developer's perspective using the NVDA build system, there is no need to worry about the Python virtual environment behind the scenes, nor should you create and or activate one manually. All NVDA build system commands will handle this transparently.
```
```suggestion
NVDA's build system commands will handle all aspects of the virtual environment.
Developers should not create or activate the virtual environment manually, unless working on the build system itself.
```

feerrenrut commented on appveyor.yml:
```diff
@@ -193,20 +203,15 @@ test_script:
-     $wc = New-Object 'System.Net.WebClient'
+     Compress-Archive -Path "$systemTestOutput\*" -DestinationPath "$testOutput\systemTestResult.zip"
+     Push-AppveyorArtifact "$testOutput\systemTestResult.zip"
+          $wc = New-Object 'System.Net.WebClient'
```
This indentation also seems wrong.
```suggestion
     $wc = New-Object 'System.Net.WebClient'
```

feerrenrut commented on readme.md:
```diff
@@ -160,9 +159,7 @@ Please see the `release` keyword argument for what compiler optimizations it wil
-Two batch files (in the root of this repository) are provided for this purpose:
-* runnvda.bat runs NVDA and blocks until NVDA is exited. Any messages written to standard out or standard error during its initialization will be written directly to the console.
-* runnvdaw.bat runs NVDA but does not block. This is the behaviour that developers are previously used to by using pythonw manually.
+To launch NVDA from source, execute runnvda.bat in the root of the repository.
```
```suggestion
To launch NVDA from source, using `cmd.exe`, execute `runnvda.bat` in the root of the repository.
```

feerrenrut reviewed (changes requested):
This is looking good! I'm happy to see that the readme for the system tests has been updated too. It actually is now easier than before to run them, that's great. 

I'm guessing that it wouldn't be so simple to make the `checkPot` stuff an external command, and even if it is, it is not really necessary for this PR. I'd be curious to hear your thoughts though, is that a direction we should head in?

michaelDCurran committed:
Apply suggestions from code review

* fix indentation.
* Re-word some documentation.

Co-authored-by: Reef Turner <feerrenrut@users.noreply.github.com>

codeofdusk commented:
PR #12133 now contains a later nvda_dmp commit than this PR. Might there be a merge conflict? Should I base that PR on this branch instead?

michaelDCurran committed:
System tests: use runnvda.bat rather than pyw source\nvda.pyw directly.

michaelDCurran committed:
scons dist: copy crt dlls from the latest installed sdk (I.e. the one that was used to compile NvDAHelper etc).

michaelDCurran committed:
appveyor: Move generation of the installed_python_packages.txt file to the very end of the appveyor run, removing the need to manually nsure and activate the Python virtual environment until then, thereby letting all the high-level commands manage it themselves, providing the same pattern a developer would use locally.

michaelDCurran committed:
Call pip freeze from a batch file in appveyor, as it seems that powershell, or at least in onFinish, the Python environement is not maintained over commands, and therefore we were fetching the global list.

feerrenrut committed:
Update buildSystemNotes

michaelDCurran committed:
buildSystemNotes: address some typos.

lukaszgo1 committed:
Remove  workaround for issue #9762 (#12142)

Latest version of Py2exe no longer invalidates signature of Python's dll so this workaround is no longer needed

lukaszgo1 committed:
Python 3.8: Mention additional system requirements for Windows 7 (#12141)

* Python 3.8: Mention that on Windows 7 KB3063858 is rewquired

* Availabe -> available

josephsl committed:
Python 3.8/gettext: replace custom pgettext implementation with built--in pgettext (#12109)

* Add-on handler/pgettext: remove pgettext namespace definition. Re #9207.

As gettext.pgettext is present, there is no need to manually assign NVDA's own pgettext function to built-in namespace for add-ons. Therefore remove this assignment.

* Language handler/pgettext: remove nVDA-specific pgettext function. re #9207.

As Python 3.8 includes gettext.pgettext, there is no need for NVDA's implementation of pgettext function. Therefore remove makePgettext and builtins namespace assignment from language handler.

* Language handler: remove unused builtins namespace import.

* Language handler/pgettext: add 'pgettext' to built-in namespace. Re #9207.

As Python 3.8 includes gettext.pgettext, add it to built-in namespace by using 'names' parameter of translation.install method.

* Update copyright header

* Revert "Add-on handler/pgettext: remove pgettext namespace definition. Re #9207."

This reverts commit 991de433b645a2a45c7319a9ca2e88c8813330a8 so add-on translations will work.

* Add-on handler/pgettext: replace NVDA-specific pgettext call with Python's pgettext function. Re #9207.

Add-on initialization will still require assigning pgettext to add-on namespace, therefore use translations.pgettext instead of now removed NVDA-specific implementation.

michaelDCurran committed:
Merge branch 'master' into py3.8

michaelDCurran committed:
Merge branch 'py3.8' into py3.8_manualVenv

feerrenrut reviewed (approved):
Looks good to me. Let's rebase this on master and merge straight to master.

feerrenrut commented:
On second thoughts, no rebase is necessary. Just make sure this is up to date with master and change the target (base) branch to master.

leonardder commented:
Let's rebase this on master and merge straight to master.

Note that #11912 is not yet part of master. Also this should be up to date with the py3.8 branch before merged into master, I assume?

michaelDCurran committed:
Update what's new

michaelDCurran committed:
appveyor: remove mention of py3.8 branch, no longer needed.
